File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/LiveRideActivity.java
package net.cyclestreets;

import net.cyclestreets.util.GPS;
import net.cyclestreets.util.MessageBox;
import net.cyclestreets.views.CycleMapView;
import net.cyclestreets.views.overlay.LiveRideOverlay;
import net.cyclestreets.views.overlay.LockScreenOnOverlay;
import net.cyclestreets.views.overlay.RouteOverlay;
import android.app.Activity;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.widget.RelativeLayout;

import com.mikepenz.iconics.context.IconicsContextWrapper;

public class LiveRideActivity extends Activity
{
  public static void launch(final Context context) {
    if (!GPS.isOn(context)) {
      MessageBox.YesNo(context,
                       "LiveRide needs the GPS location service.\n\nWould you like to turn it on now?",
                       new DialogInterface.OnClickListener() {
                          public void onClick(DialogInterface arg0, int arg1) {
                            GPS.showSettings(context);
                          }
                       });
      return;
    }
    launchActivity(context);
  }

  private static void launchActivity(final Context context) {
    final Intent intent = new Intent(context, LiveRideActivity.class);
    context.startActivity(intent);
  }

  private CycleMapView map_;

  @Override
  protected void attachBaseContext(Context newBase) {
    // Allows the use of Material icon library, see https://github.com/mikepenz/Android-Iconics
    super.attachBaseContext(IconicsContextWrapper.wrap(newBase));
  }

  @Override
  public void onCreate(final Bundle saved) {
    super.onCreate(saved);

    // Map initialized in onResume
  }

  //////////////////////////
  @Override
  public void onPause() {
    map_.disableFollowLocation();
    map_.onPause();

    super.onPause();
  }

  @Override
  public void onResume() {
    super.onResume();

    // Map needs to be recreated, because tile provider is shut down on CycleMapView.onPause
    initializeMapView();
    map_.onResume();
    map_.enableAndFollowLocation();
  }

  private void initializeMapView() {
    map_ = new CycleMapView(this, this.getClass().getName());
    map_.overlayPushBottom(new RouteOverlay());
    map_.overlayPushTop(new LockScreenOnOverlay(this, map_));
    map_.overlayPushTop(new LiveRideOverlay(this, map_));
    map_.lockOnLocation();
    map_.hideLocationButton();

    final RelativeLayout rl = new RelativeLayout(this);
    rl.addView(map_, new RelativeLayout.LayoutParams(RelativeLayout.LayoutParams.MATCH_PARENT, RelativeLayout.LayoutParams.MATCH_PARENT));
    setContentView(rl);
  }

}

File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/LiveRideActivity.kt
package net.cyclestreets

import android.Manifest
import net.cyclestreets.liveride.LiveRideService
import net.cyclestreets.util.GPS
import net.cyclestreets.util.MessageBox
import net.cyclestreets.views.CycleMapView

import android.app.Activity
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.location.Location
import android.os.Bundle
import android.os.IBinder
import android.util.Log
import android.widget.RelativeLayout

import net.cyclestreets.util.Logging
import net.cyclestreets.util.hasPermission
import net.cyclestreets.views.overlay.*


private val TAG = Logging.getTag(LiveRideActivity::class.java)


class LiveRideActivity : Activity(), ServiceConnection, LiveRideOverlay.Locator {
    private lateinit var map: CycleMapView
    private lateinit var liveride: LiveRideService.Binding

    override fun onServiceConnected(className: ComponentName, binder: IBinder) {
        liveride = binder as LiveRideService.Binding

        if (!liveride.areRiding())
            liveride.startRiding()
    }

    override fun onServiceDisconnected(className: ComponentName) {}

    override fun lastLocation(): Location? {
        return liveride.lastLocation()
    }

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val intent = Intent(this, LiveRideService::class.java)
        this.bindService(intent, this, Context.BIND_AUTO_CREATE)
    }

    public override fun onDestroy() {
        if (this::liveride.isInitialized) {
            liveride.stopRiding()
        }
        this.unbindService(this)
        super.onDestroy()
    }

    public override fun onPause() {
        map.disableFollowLocation()
        map.onPause()
        super.onPause()
    }

    public override fun onResume() {
        super.onResume()

        // Map needs to be recreated, because tile provider is shut down on CycleMapView.onPause
        initializeMapView()
        map.onResume()
        map.enableAndFollowLocation()
    }

    private fun initializeMapView() {
        map = CycleMapView(this, this.javaClass.name).apply {
            overlayPushBottom(RouteOverlay())
            overlayPushTop(WaymarkOverlay(this))
            overlayPushTop(LockScreenOnOverlay(this))
            overlayPushTop(RotateMapOverlay(this))
            overlayPushTop(LiveRideOverlay(this@LiveRideActivity, this@LiveRideActivity))
            lockOnLocation()
            hideLocationButton()
            shiftAttribution()
        }
        RelativeLayout(this).apply {
            addView(map,
                    RelativeLayout.LayoutParams(RelativeLayout.LayoutParams.MATCH_PARENT,
                                                RelativeLayout.LayoutParams.MATCH_PARENT))
            this@LiveRideActivity.setContentView(this)
        }
    }

    companion object {
        fun launch(context: Context) {
            if (!hasPermission(context, Manifest.permission.ACCESS_FINE_LOCATION)) {
                // Should be unreachable but we're being defensive
                Log.w(TAG, "Location permission is not granted.  Bail out.")
                return
            }

            if (!GPS.isOn(context)) {
                MessageBox.YesNo(context,
                    "LiveRide needs the GPS location service.\n\nWould you like to turn it on now?")
                { _, _ -> GPS.showSettings(context) }
                return
            }

            // Proceed
            context.startActivity(Intent(context, LiveRideActivity::class.java))
        }
    }
}
